lib/attr_extras/attr_initialize.rb

```diff
@@ -14,20 +16,21 @@ def apply
     validate_arity = method(:validate_arity)
     set_ivar_from_hash = method(:set_ivar_from_hash)
 
+    declaration = AttributesDeclaration.new(names)
+
     klass.send(:define_method, :initialize) do |*values|
       validate_arity.call(values.length, self.class)
 
-      names.zip(values).each do |name_or_names, value|
-        if name_or_names.is_a?(Array)
-          hash = value || {}
-
-          name_or_names.each do |name|
-            set_ivar_from_hash.call(self, name, hash)
-          end
-        else
-          name = name_or_names
-          instance_variable_set("@#{name}", value)
-        end
+      positional_attributes_values = values.take(declaration.positional_attributes.count)
+      hash_attributes_values = values.drop(declaration.positional_attributes.count)
+
+      declaration.positional_attributes.zip(positional_attributes_values).each do |name, value|
+        instance_variable_set("@#{name}", value)
+      end
+
+      hash_values_with_defaults = declaration.default_values.merge(hash_attributes_values.first || {})
```

Comment:
Having an array of a single item here reads a bit odd. Could it make sense to do something like `hash_attributes_values = values.drop(declaration.positional_attributes.count).first` above?